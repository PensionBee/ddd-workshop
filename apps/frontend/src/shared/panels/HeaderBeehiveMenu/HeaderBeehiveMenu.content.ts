import type { NavMenuItem } from "@shared/components/Nav/Nav.types";

type HeaderBeehiveMenuContentProps = {
  beehiveMenu?: {
    canAccessBalance?: boolean;
    canAccessAnalytics?: boolean;
    canAccessContributions?: boolean;
    canAccessVerifyDetails?: boolean;
    canAccessTransfers?: boolean;
    canAccessWithdrawals?: boolean;
    canAccessInvestmentPathways?: boolean;
    canAccessPensionReport?: boolean;
    canAccessPlan?: boolean;
  };
};

export const getDynamicHeaderBeehiveMenuContent = (
  menuPermissions?: HeaderBeehiveMenuContentProps
) => {
  return [
    {
      children: "Balance",
      href: "/beehive/dashboard",
      hidden: !menuPermissions?.beehiveMenu?.canAccessBalance,
    },
    {
      children: "Analytics",
      href: "/beehive/dashboard/analytics",
      hidden: !menuPermissions?.beehiveMenu?.canAccessAnalytics,
    },
    {
      children: "Contributions",
      href: "/beehive/dashboard/contributions",
      hidden: !menuPermissions?.beehiveMenu?.canAccessContributions,
    },
    { children: "Add a pension", href: "/beehive/dashboard/add" },
    {
      children: "Verify my details",
      href: "/beehive/dashboard/id",
      hidden: !menuPermissions?.beehiveMenu?.canAccessVerifyDetails,
    },
    {
      children: "Transfers",
      href: "/beehive/dashboard/tracker",
      hidden: !menuPermissions?.beehiveMenu?.canAccessTransfers,
    },
    {
      children: "Withdrawals",
      href: "/beehive/dashboard/withdrawals",
      hidden: !menuPermissions?.beehiveMenu?.canAccessWithdrawals,
    },
    {
      children: "Investment pathways",
      href: "/beehive/dashboard/investment-pathways",
      hidden: !menuPermissions?.beehiveMenu?.canAccessInvestmentPathways,
    },
    {
      children: "Detailed pension report",
      href: "/beehive/dashboard/report",
      hidden: !menuPermissions?.beehiveMenu?.canAccessPensionReport,
    },
    { children: "Support", href: "/beehive/dashboard/help-support" },
    {
      children: "Account",
      subList: [
        {
          children: "My Plan",
          href: "/beehive/dashboard/my-plan",
          hidden: !menuPermissions?.beehiveMenu?.canAccessPlan,
        },
        { children: "My Profile", href: "/beehive/dashboard/profile" },
        { children: "Beneficiaries", href: "/beehive/dashboard/beneficiaries" },
        {
          children: "Documents & Resources",
          href: "/beehive/dashboard/documents-resources",
        },
        {
          children: "Two-factor authentication",
          href: "/beehive/dashboard/two-factor",
        },
        {
          children: "Refer a friend",
          href: "/beehive/dashboard/rewards",
          icon: "Heart",
        },
      ],
    },
  ] satisfies NavMenuItem[];
};
