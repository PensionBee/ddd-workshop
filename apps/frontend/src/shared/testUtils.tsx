import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
