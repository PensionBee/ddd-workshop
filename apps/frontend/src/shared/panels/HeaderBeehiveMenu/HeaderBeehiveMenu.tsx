import { usePermissionsQuery } from "@modules/_app/api/appQueries";
import { GridContainer } from "@shared/components/Grid";
import { AccordionMenu } from "@shared/components/Nav/AccordionMenu/AccordionMenu";
import { MainNav } from "@shared/components/Nav/MainNav/MainNav";
import { useMediaQueryLg } from "@shared/hooks/useMediaQuery/useMediaQuery";
import { Box } from "@shared/components/Box";
import { getDynamicHeaderBeehiveMenuContent } from "./HeaderBeehiveMenu.content";

export const HeaderBeehiveMenu: React.FC = (props) => {
  const isDesktop = useMediaQueryLg();
  const { isLoading, data: permissions } = usePermissionsQuery();

  const dynamicBeehiveMenuList = getDynamicHeaderBeehiveMenuContent(
    permissions || {}
  );

  if (isLoading)
    return <Box h={{ xs: 55, lg: 47 }} mt={{ xs: 65, lg: 100 }}></Box>;

  return (
    <GridContainer
      data-rc="HeaderBeehiveMenu"
      {...props}
      data-state={isDesktop ? "desktop" : "mobile"}
      pt={16}
    >
      {isDesktop ? (
        <MainNav
          list={dynamicBeehiveMenuList}
          colGapLarge={25}
          borderBottom
          isTabMenu
        />
      ) : (
        <AccordionMenu list={dynamicBeehiveMenuList} />
      )}
    </GridContainer>
  );
};
