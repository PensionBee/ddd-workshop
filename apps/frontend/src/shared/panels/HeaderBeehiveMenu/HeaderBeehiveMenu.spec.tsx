import { mockMediaMatches, routerRender, vi } from "@shared/testUtils";
import { HeaderBeehiveMenu } from "./HeaderBeehiveMenu";
import { getDynamicHeaderBeehiveMenuContent } from "./HeaderBeehiveMenu.content";

vi.mock("@modules/_app/api/appQueries", () => ({
  _esModule: true,
  usePermissionsQuery: () => ({
    isLoading: false,
    data: {
      beehiveMenu: {
        canAccessBalance: false,
      },
    },
  }),
}));

describe("HeaderBeehiveMenu", () => {
  const TEST_ID = "HeaderBeehiveMenu";
  it("should render on desktop", () => {
    mockMediaMatches("lg");
    const { getByTestId, asFragment } = routerRender(
      <HeaderBeehiveMenu data-testid={TEST_ID} />
    );
    const menu = getByTestId(TEST_ID);
    expect(menu).toBeInTheDocument();
    expect(menu).toHaveAttribute("data-state", "desktop");
    expect(asFragment).toMatchSnapshot();
  });

  it("should render on mobile", () => {
    mockMediaMatches("xs");
    const { getByTestId, asFragment } = routerRender(
      <HeaderBeehiveMenu data-testid={TEST_ID} />
    );
    const menu = getByTestId(TEST_ID);
    expect(menu).toBeInTheDocument();
    expect(menu).toHaveAttribute("data-state", "mobile");
    expect(asFragment).toMatchSnapshot();
  });
});

describe("HeaderBeehiveMenuContent", () => {
  describe("Default", () => {
    it("should hide hidden links", () => {
      getDynamicHeaderBeehiveMenuContent().forEach((item) => {
        if (Object.prototype.hasOwnProperty.call(item, "hidden")) {
          return expect(item.hidden).toBe(true);
        }
        return expect(item.hidden).toBe(undefined);
      });
    });
  });

  describe("Dynamic", () => {
    it("should hide hidden links", () => {
      getDynamicHeaderBeehiveMenuContent().forEach((item) => {
        if (Object.prototype.hasOwnProperty.call(item, "hidden")) {
          return expect(item.hidden).toBe(true);
        }
        return expect(item.hidden).toBe(undefined);
      });
    });

    it("should show links when permissions allow it", () => {
      getDynamicHeaderBeehiveMenuContent({
        beehiveMenu: {
          canAccessBalance: true,
        },
      }).forEach((item) => {
        if (item.children === "Balance") {
          return expect(item.hidden).toBe(false);
        }
        if (Object.prototype.hasOwnProperty.call(item, "hidden")) {
          return expect(item.hidden).toBe(true);
        }
        return expect(item.hidden).toBe(undefined);
      });
    });
  });
});
