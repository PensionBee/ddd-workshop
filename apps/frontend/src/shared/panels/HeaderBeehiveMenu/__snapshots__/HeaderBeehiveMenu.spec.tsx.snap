// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderBeehiveMenu should render on desktop 1`] = `[Function]`;

exports[`HeaderBeehiveMenu should render on mobile 1`] = `[Function]`;
